File: C++/Solution200.cpp
#include <iostream>
#include <vector>

#include <unordered_map>
#include <unordered_set>

using namespace std;

class Solution200
{
public:
    int ans = 0;
    int n, m;

    int xx[4] = {0, 1, 0, -1};
    int yy[4] = {1, 0, -1, 0};

    void dfs(vector<vector<char>> &grid, int x, int y)
    {
        grid[x][y] = '0';
        for (int i = 0; i < 4; i++)
        {
            int curx = x + xx[i], cury = y + yy[i];
            if (curx >= 0 && curx < n && cury >= 0 && cury < m && grid[curx][cury] == '1')
            {
                dfs(grid, curx, cury);
            }
        }
    }

    int numIslands(vector<vector<char>> &grid)
    {
        n = grid.size();
        m = grid[0].size();
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (grid[i][j] == '1')
                {
                    ans++;
                    dfs(grid, i, j);
                }
            }
        }
        return ans;
    }
};

File: Java/other/Solution200.java
import java.util.Set;
import java.util.HashSet;

class Solution200 {
	int number = 0;
	Set<Integer> set = new HashSet<>();
	int[] x = new int[]{0, 0, -1, 1};
	int[] y = new int[]{-1, 1, 0, 0};

    public int numIslands(char[][] grid) {
    	if(grid.length == 0){
    		return number;
    	}
    	for(int i = 0; i < grid.length; i++){
    		for(int j = 0; j < grid[0].length; j++){
    			if(grid[i][j] == '1' && !set.contains(i * grid[0].length + j)){
    				number += 1;
    				dfs(grid, i, j);
    			}
    		}
    	}
    	return number;
    }

    private void dfs(char[][] grid, int row, int col){
    	set.add(row * grid[0].length + col);
    	for(int index = 0; index < 4; index++){
    		if(row + x[index] >= 0 && row + x[index] < grid.length && col + y[index] >= 0 && col + y[index] < grid[0].length && grid[row + x[index]][col + y[index]] == '1'){
    			int cur = (row + x[index]) * grid[0].length + col + y[index];
    			if(!set.contains(cur)){
    			set.add(cur);
    			dfs(grid, row + x[index], col + y[index]);
    		}
    		}
    	}
    }

}